packages/swingset-liveslots/src/virtualObjectManager.js

```diff
@@ -754,7 +754,7 @@ export function makeVirtualObjectManager(
       }
       case 'many': {
         assert(multifaceted);
-        facetNames = Object.getOwnPropertyNames(behavior).sort();
+        facetNames = ownKeys(behavior).sort();
```

Comment:
Not technically just a refactor, since `ownKeys` will enumerate non-enumerable properties as well as string-named properties. I think this is no less correct because `assessFacetiousness` also uses `ownKeys`. But reviewers should judge for themselves.

Reply:
I had to check the unit tests to remind myself that facet *names* may not be symbols, but the methods themselves may use symbols as properties as long as their values are functions. `assessFacetiousness` does a one-level recursion of the behavior record, and it is important that the top level examine all properties (string- and symbol- named) so it can correctly reject symbol-named facets. So we definitely can't use `getOwnPropertyNames` in `assessFacetiousness`: we must use `Reflect.ownKeys()` there.

Here, on line 757, we've already run `assessFacetiousness`, and it judged the behavior record as `many`, which means we know that all the facet names are strings. So switching from `getOwnPropertyNames` to `ownKeys` won't surprise us with symbol-named properties. And both `getOwnPropertyNames` and `Reflect.ownKeys` return non-enumerable properties, so I agree it is no less correct.

But...

when doing that analysis just now, I briefly forgot that `getOwnPropertyNames` sees non-enumerable properties. So I figured I should think about what happens if that were actually a change in behavior.

If the behavior record included a non-enumerable string-named facet:
```js
const behavior = {
  incrementer: { incr: ({state}) => state.count += 1 },
};
Object.defineProperty(behavior, 'resetter', {
  enumerable: false,
  value: { reset: ({state}) => state.count = 0 },
});
```

.. well, I had to write a test to find out. It turns out that we don't entirely ignore the "hidden" facet: we create a cohort that includes both `incrementer` and `resetter`. But the `resetter` facet is not populated with methods: it is a plain empty object, whose prototype is the plain `Object.prototype`, not a special method-bearing `defendPrototype` construction.

Which is weird. We shouldn't be doing that: we should either ignore the facet completely, or we should create a functional facet, not something in between.

`defendPrototypeKit` starts out by extracting the `facetNames` with (Reflect.)`ownKeys`:

https://github.com/endojs/endo/blob/4c8702a77923c4cf757718296d34d933b30b63a4/packages/exo/src/exo-tools.js#L277

but then it uses `objectMap` to enumerate all the facets and build a record of per-facet prototypes:

https://github.com/endojs/endo/blob/4c8702a77923c4cf757718296d34d933b30b63a4/packages/exo/src/exo-tools.js#L295-L302

and `objectMap` uses `entries`:

https://github.com/endojs/endo/blob/4c8702a77923c4cf757718296d34d933b30b63a4/packages/patterns/src/utils.js#L100-L104

which is `Object.entries`:

https://github.com/endojs/endo/blob/4c8702a77923c4cf757718296d34d933b30b63a4/packages/patterns/src/utils.js#L7

which returns only enumerable string-named properties. So we get `prototypeKit.resetter = undefined`, which causes the facet to get the normal `Object.prototype` instead of the special behavior-bearing one.

I see two options to make this more rational:
* 1: change the VOM to ignore non-enumerable facets (honor the user's possibly-misguided attempt to hide them)
  * change line 757 to use `Object.keys`, which only sees *enumerable* string-named properties
  * `assessFacetiousness` will still complain if a non-enumerable property violates the rules, but if the record passes the facetiousness check, then `facetNames` will ignore them, and we won't attempt to make facets for them
  * `defendPrototypeKit` will internally build a larger `facetNames` list than the VOM does, but its use of `objectMap` means the `prototypeKit` it returns will not
* 2: change Endo's `defendPrototypeKit` to process the non-enumerable facets (i.e. consistently ignore the user's attempts to hide them)
  * use something other than `objectMap` to build the list, probably a loop over its internal `facetNames` list

@erights @FUDCo thoughts? I'm kind of inclined to pick "1" (`Object.keys`), but mostly because it avoids another Endo release cycle.

I'll add a branch with a test case.

Reply:
created https://github.com/Agoric/agoric-sdk/issues/7446 for this particular issue

Reply:
Thanks for checking! A lot to ponder here. Putting this into Draft until resolved.

Reply:
3: For the record of facet names, adapt the same property restrictions as copyRecord: Each property must be a
   * string named
   * enumerable
   * data property
And the object itself must inherit directly from `Object.prototype`.

As explained at https://github.com/endojs/endo/blob/master/packages/pass-style/doc/copyRecord-guarantees.md , by imposing these restrictions, we ensure that all the property enumerating operations at https://github.com/endojs/endo/blob/master/packages/pass-style/doc/copyRecord-guarantees.md#how-do-i-enumerate-thee-let-me-list-the-ways give the same answers, preventing hazards like my `objectMap` mistake from being mistaken.

   * `Object.keys`, `Object.values`, `Object.entries`, `{ ... }`, `[...]`, and all but the first argument to `Object.assign`.
      * Only string-named enumerable own. But does a GET on accessor properties.
   * `Reflect.ownKeys`
      * all own property names. Nothing inherited
   * `Object.getOwnPropertyNames`
      * own, string-named, whether or not enumerable
   * `Object.getOwnPropertyDescriptors`
      * own, whether named by string or symbol, whether or not enumerable
   * `for/in` loop (thankfully banned by eslint)
      * all enumerable string named, whether own or inherited.

This restriction must be enforced on the endo side, which I can do. However, as long as the code in agoric-sdk stays within this restriction, as it has sanely done, imposing the restriction won't actually make any operational difference. Thus, we need not wait for an endo release cycle for agoric-sdk to proceed assuming the restriction.

Finally, imposing the restriction is, I would argue, makes the most sense semantically because the kit being made is itself a copyRecord of remotables. So the emitted facet names must all obey these restrictions anyway.

How does (3) sound?

Reply:
I propose that we enforce the same restriction on the state record returned by the `init` function.